src/manager.rs

```diff
@@ -0,0 +1,132 @@
+        if let Err(_) = fs::metadata(&path) {
+            try!(fs::create_dir(&path));
+        }
+
+        Ok(ScriptManager {
+          path: path.clone(),
+          phantom: PhantomData,
+          executions: HashMap::new(),
+          api: api
+        })
+    }
+
+    /// Attempt to read, parse and launch the scripts at the given path, asynchronously.
+    ///
+    /// If any script causes an error, report that script, but do not let that error prevent other scripts from being read/parsed/launched.
+    ///
+    /// Only scripts that are enabled should be loaded.
+    pub fn load<F>(&self, cb: F) where F: FnOnce(HashMap<String, Result<(), Error>>) + Send {
+        let (tx, rx) = channel();
+        let mut launching_count = 0;
+
+        for entry in try!(fs::read_dir(&self.path)) {
+            let path = try!(entry).path();
+            let id = path.file_stem().unwrap().to_str().unwrap(); // strips the file extension
+
+            // Load the script from this file.
+            let f = try!(File::open(&path));
+            let content = String::new();
+            try!(f.read_to_string(&mut content));
+            let script = try!(Parser::parse(content));
+
+            launching_count += 1;
+            self.launch_script(id.to_owned(), script, move |result| {
+                tx.send((id, result)).unwrap();
+            });
+        }
+
+        // Wait for the results to come in.
+        thread::spawn(move || {
+            let result_map = HashMap::new();
+            while launching_count > 0 {
+                let (id, result) = rx.recv().unwrap();
+                result_map.insert(id.to_owned(), result);
+                launching_count -= 1;
+            }
+            cb(result_map);
+        });
+    }
+
+    fn launch_script<F>(&self, id: String, script: Script<UncheckedCtx>, cb: F)
+            where F: FnOnce(Result<(), Error>) + Send + 'static {
+
+        let mut runner = Execution::<Env>::new();
+        let (tx, rx) = channel();
+        let executions = self.executions;
+        runner.start(self.api, script, move |result| {
+            if let ExecutionEvent::Starting { result: Ok(()) } = result {
+                executions.insert(id, runner);
+                cb(Ok(()));
+            } else {
+                cb(Err(Error::new(ErrorKind::InvalidData,
+                                  format!("Unable to launch script: {}", id))));
+            }
+        });
+    }
+
+    fn save_script(&self, id: String, source: String) {
+
+    }
+
+    /// Attempt to add a new script, asynchronously.
+    ///
+    /// If the script can be parsed/compiled/executed, this script is immediately saved to the disk.
+    fn put<F>(&self, id: &String, source: &String, cb: F) where F: FnOnce(Result<(), Error>) + Send {
+        match Parser::parse(source.to_owned()) {
+            Ok(script) => {
+                self.launch_script(id.to_owned(), script, move |result| {
```

Comment:
Oh, right, calling `self` in a `move` callback is seldom a good idea. I have essentially two workarounds for this kind of situation:

- using static methods instead of dynamic methods (so no `&self`);
- replacing the method call by a message dispatched to a thread (typically, the same thread).